### ISSUE TITLE
Index Nested Loop Join

### ISSUE THREAD

**Issue by Mytherin** [cwida/duckdb#115]
For cases where we know the RHS is much smaller than the LHS of a join, and we have an index on the join key, we can do an index nested loop join. In the index nested loop join you use the keys on the RHS to do an index lookup into the LHS. As we are using unclustered indexes, this is generally only worth it if the RHS is very small (e.g. less than 0.1% of the LHS).

This is the case though if e.g. we have an IN() list that has been folded into a join; we know the exact cardinality of the IN() list, hence we can reliably determine if it is worth it or not.

**Comment by aris-koning** [comment]
Are you guys in a hurry with this one? If not can I pickt it up?

**Comment by pholanda** [comment]
Hey, @aris-koning I'm currently developing an ART as a default secondary index, but I got a bit stuck because of a VLDB revision I got. I guess this task is dependent on the actual existence of indexes, but there is a (pretty bad) order index implementation there that you can use as a basis there.

**Comment by aris-koning** [comment]
I am not in a hurry myself and I can take a look at the order index. With ART you mean Adaptive Radix Tree? Is your ongoing work in a branch or fork somewhere? It would already be helpful if these indices are going to have a common interface layout.

**Comment by pholanda** [comment]
Yeah, all indexes will have the same interface and Art is the adaptive radix tree. 

Tbh, I guess the plan is to actually drop the order index (I only did it to have a feeling on how to actually implement the indexes in the system) and replace it with a bw-tree or b+tree in the near future since (I think) they are more efficient for range queries than an ART.

@Mytherin can correct me if I'm wrong, but I think for this task you just need to change the index optimizer layer, but I only have it in the ART branch as well.

**Comment by Mytherin** [comment]
For this task you need to create a new join operator, IndexNestedLoopJoin, that would replace a scan that would be fed into the join on the LHS with index lookups based on the values obtained from the RHS. I think you can internally use the PhysicalIndexScan to perform the lookups on the LHS inside the IndexNestedLoopJoin implementation. In that way, I think any future changes to the interface shouldn't break anything too much as any changes to the index will also update the PhysicalIndexScan to work with the index (obviously).

To answer your other question @aris-koning this is not a high priority issue, it is fine if you take it, I think you can work on it in parallel to @pholanda working on the actual index without conflicts.
